lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,546 @@
-            // the predecessors (the visiting set).
-            if (visiting.contains(childVal)) {
-              // Comb Cycle Detected !!
-              reportLoopFound(childVal, visiting);
-              return failure();
-            }
-          }
-          // child nodes added, continue the DFS
-          if (dfsSize != dfsStack.size())
-            continue;
-        }
-        // FieldRef is an SCC root node, pop the visiting stack to remove the
-        // nodes that are no longer active predecessors, that is their sub-tree
-        // is already explored. All the Values reachable from `dfsVal` have been
-        // explored, remove it and its children from the visiting stack.
-        visiting.popUntilVal(dfsVal);
-
-        auto popped = dfsStack.pop_back_val();
-        (void)popped;
-        LLVM_DEBUG({
-          llvm::dbgs() << "\n dfs popped :"
-                       << getFieldName(FieldRef(popped, 0)).first;
-          dump();
-        });
-      }
-    }
-
-    return success();
+    constructReachingDefGraph(module);
+    return dfsTraverse(reachingDefGraph);
   }
 
-  // Preprocess the module ops to get the
-  // 1. roots for DFS traversal,
-  // 2. FieldRef corresponding to each Value.
-  void preprocess(SmallVector<Value> &worklist) {
-    // All the input ports are added to the worklist.
-    for (BlockArgument arg : module.getArguments()) {
-      auto argType = type_cast<FIRRTLType>(arg.getType());
-      if (type_isa<RefType>(argType))
+  void constructReachingDefGraph(FModuleOp module) {
+    LLVM_DEBUG(llvm::dbgs() << "\n Module :" << module.getName());
+
+    // ALl the module output ports must be added as the initial nodes to the
+    // reaching def graph.
+    for (auto port : module.getArguments()) {
+      if (module.getPortDirection(port.getArgNumber()) != Direction::Out)
         continue;
-      if (module.getPortDirection(arg.getArgNumber()) == Direction::In)
-        worklist.push_back(arg);
-      if (!argType.isGround())
-        setValRefsTo(arg, FieldRef(arg, 0));
+      auto portNode = getOrAddNode(port);
+      auto portField = reachingDefGraph[portNode].first;
+      // If aggregate type, then add all the ground type fields to the graph.
+      if (auto bType = getBaseType(port.getType()))
+        if (!bType.isGround())
+          walkGroundTypes(bType, [&](uint64_t index, FIRRTLBaseType t) {
+            getOrAddNode(portField.getSubField(index));
+          });
     }
-    DenseSet<Value> memPorts;
-
-    for (auto &op : module.getOps()) {
-      TypeSwitch<Operation *>(&op)
-          // Wire is added to the worklist
-          .Case<WireOp>([&](WireOp wire) {
-            worklist.push_back(wire.getResult());
-            auto ty = type_dyn_cast<FIRRTLBaseType>(wire.getResult().getType());
-            if (ty && !ty.isGround())
-              setValRefsTo(wire.getResult(), FieldRef(wire.getResult(), 0));
+    for (auto &op : module) {
```

Comment:
I think it's necessary to walk the operations in the presence of `Group`.

Reply:
Thanks, fixed it.